README.md
# react-simple-video-player
A simple, material design video player component for React.

![Video player screenshot](https://github.com/rafaelklaessen/react-simple-video-player/raw/master/screenshots/screenshot.png "Screenshot of the video player")

## Install
`yarn add react-simple-video-player` or `npm install --save react-simple-video-player`

## Usage
### Basic example
```javascript
import VideoPlayer from 'react-simple-video-player';

const App = () => (
  <VideoPlayer url="/video.mp4" />
);
```

### Advanced example
```javascript
import VideoPlayer from 'react-simple-video-player';

const App = () => (
  <VideoPlayer
    url="/video.mp4"
    poster="/myPoster.png"
    width={400}
    height={300}
    autoplay
  />
);
```

## Props
`*` = Required

Prop | Description | default
---- | ----------- | -------
`url*`| Url of the video file to play |
`poster` | Url of the image to use as poster for the video |
`width` | Width of the video player | `640`
`height` | Height of the video player | `360`
`autosize` | If set to true, the video player will become 100% of the parent width and 100% of the parent height (if the parent does not have a height, the video player will automatically become the right height). `autosize` overrides the given `width` and `height` | `false`
`autoplay` | If set to true, the video will automatically play | `false`
`loop` | If set to true, the video will loop | `false`
`aspectRatio` | The aspect ratio of the video (see explanation below). | `16:9`

#### Usage of the `aspectRatio` prop
The `aspectRatio` prop can be used to automatically give the video player the correct height for the given width (or the correct width for the given height).

For example, in the following scenario:
```javascript
<VideoPlayer url="/video.mp4" width={400} aspectRatio="4:3" />
```
the video player would automatically become 300px high.

In the following scenario:
```javascript
<VideoPlayer url="/video.mp4" height={300} aspectRatio="4:3" />
```
the video player would automatically become 400px wide.

In the following scenario:
```javascript
<VideoPlayer url="/video.mp4" aspectRatio="4:3" />
```
the video player would automatically become 480px high (since the width defaults to 640px).

Notice that when an invalid `aspectRatio` is given, it'll be ignored.

## Notes
`react-simple-video-player` is built on top of `react-player`.
#### This project does *not* aim to provide a customisable video player.
Please use [react-player](https://www.npmjs.com/package/react-player) when you need customisability.


CHANGELOG.md
# Version 1.0.0
Released 2018-02-03

- Updated README
- Added support for video looping

# Version 0.2.2
Released 2018-02-02

- Removed glamorous support because it causes duplicate glamor instances

# Version 0.2.1
Released 2018-02-02

- Added glamorous support

# Version 0.2.0
Released 2018-02-02

- Added tests

# Version 0.1.1
Released 2018-02-02

- Added README content

# Version 0.1.0
Released 2018-02-02

- Initial release


src/index.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import VideoPlayer from './index';

describe('<VideoPlayer />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<VideoPlayer />, div);
  });
});


src/icons/__tests__/PlayIcon.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import PlayIcon from '../PlayIcon';

describe('<PlayIcon />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<PlayIcon />, div);
  });
});


src/icons/__tests__/PauseIcon.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import PauseIcon from '../PauseIcon';

describe('<PauseIcon />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<PauseIcon />, div);
  });
});


src/index.js
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import SizedPlayer from './SizedPlayer';
import PlayIcon from './icons/PlayIcon';
import PauseIcon from './icons/PauseIcon';

const Player = glamorous.div({
  position: 'relative',
  lineHeight: 0,
  display: 'inline-block'
}, ({ autosize }) => {
  if (autosize) return { width: '100%', height: '100%' };
});

const overlayStyles = {
  position: 'absolute',
  top: 0,
  left: 0,
  width: '100%',
  height: '100%'
};

const Overlay = glamorous.div(overlayStyles, {
  backgroundColor: 'rgba(0, 0, 0, .5)',
  opacity: 0,
  transition: 'all 450ms cubic-bezier(.23, 1, .32, 1) 0ms',
  cursor: 'pointer'
}, ({ shown }) => {
  if (shown) return { opacity: 1 };
});

const InnerOverlay = glamorous.div(overlayStyles);

const PlayButton = glamorous.div({
  position: 'absolute',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: 72,
  height: 72
});

export default class VideoPlayer extends PureComponent {
  static propTypes = {
    url: PropTypes.string.isRequired,
    poster: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    autosize: PropTypes.bool,
    autoplay: PropTypes.bool,
    loop: PropTypes.bool,
    aspectRatio: PropTypes.string
  };

  static defaultProps = {
    loop: false,
    autoplay: false
  };

  state = {
    playing: this.props.autoplay,
    overlayShown: !this.props.autoplay
  };

  onPlay = () => {
    this.setState({
      playing: true,
      overlayShown: false
    });
  };

  onPause = () => {
    this.setState({
      playing: false,
      overlayShown: true
    });
  };

  onMouseOver = () => this.setState({ overlayShown: true });

  onMouseLeave = () => {
    if (!this.state.playing) {
      this.setState({ overlayShown: true });
      return;
    }

    this.setState({ overlayShown: false });
  };

  render = () => {
    const playing = this.state.playing;
    const autosize = this.props.autosize;
    return (
      <Player
        autosize={autosize}
        onMouseOver={this.onMouseOver}
        onMouseLeave={this.onMouseLeave}
      >
        <SizedPlayer playing={playing} {...this.props} />
        <Overlay
          shown={this.state.overlayShown}
          onClick={playing ? this.onPause : this.onPlay}
        >
          <PlayButton>
            {playing ? <PauseIcon /> : <PlayIcon />}
          </PlayButton>
          <InnerOverlay />
        </Overlay>
      </Player>
    );
  }
}


src/SizedPlayer.js
import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import ReactPlayer from 'react-player';

const PlayerContainer = glamorous.div(({ width, height, autosize }) => {
  if (autosize) return { width: '100%', height: '100%' };
  return { width, height };
});

export const parseAspectRatio = (aspectRatio) => {
  const ratioArr = aspectRatio.split(':');
  const ratio = ratioArr.map(str => Number(str));
  if (ratio.length !== 2 || Number.isNaN(ratio[0]) || Number.isNaN(ratio[1])) {
    // eslint-disable-next-line
    console.warn('Invalid aspect ratio supplied to react-simple-video-player, defaulting to 16:9');
    return [16, 9];
  }
  return ratio;
};

const SizedPlayer = ({ url, poster, playing, loop, ...props }) => {
  const { ...options } = props;
  const aspectRatio = parseAspectRatio(options.aspectRatio);

  if (!options.autosize && !options.width && !options.height) {
    options.width = 640;
    options.height = options.width / aspectRatio[0] * aspectRatio[1];
  }

  if (!options.autosize && options.width && !options.height) {
    options.height = options.width / aspectRatio[0] * aspectRatio[1];
  }

  if (!options.autosize && options.height && !options.width) {
    options.width = options.height / aspectRatio[1] * aspectRatio[0];
  }

  const config = poster ? { file: { attributes: { poster } } } : {};

  return (
    <PlayerContainer
      width={options.width}
      height={options.height}
      autosize={options.autosize}
    >
      <ReactPlayer
        url={url}
        config={config}
        playing={playing}
        loop={loop}
        width="100%"
        height="100%"
      />
    </PlayerContainer>
  );
};

SizedPlayer.propTypes = {
  url: PropTypes.string.isRequired,
  poster: PropTypes.string,
  playing: PropTypes.bool.isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  autosize: PropTypes.bool,
  loop: PropTypes.bool,
  aspectRatio: PropTypes.string
};

SizedPlayer.defaultProps = {
  autosize: false,
  aspectRatio: '16:9'
};

export default SizedPlayer;
